018-function-execution-model-refactor/018-function-execution-model-refactor.md

```diff
@@ -0,0 +1,129 @@
+
+The executor will have multiple tasks but the main one is to spin up runtimes when they are needed and to update their code whenever functions are updated. Runtimes will now be web servers based of their relevant languages and when code is added it will be added to the Runtime web server's router.
+
+The executor will also act like a router itself, when API Requests that are for functions are received it will be the executor's job to correlate them to the relevant runtimes and forward them for processing. When the Runtime's web server is done processing the request and returns a response the executor will return the response to whatever client requested it.
+
+#### Scaling
+Scaling this solution will need to be worked on in the RFC but a solution I thought of would be to have the executor to keep track of resources being used by runtimes and the total resources being used hits a certain threshold another instance of the runtime will be spun up and requests will be load balanced onto the multiple runtime instances.
+
+A separate RFC is currently being worked on for how runtime languages will be written. 
+
+#### API Endpoints
+This rewrite will add functionality to the API endpoints that are currently used by the functions worker. The executor will take over these endpoints while the rest will stay underneath the functions worker.
+
+`POST /v1/functions/{id}/execute` - Execute function API endpoint
+The standard execution endpoint for functions will be rewritten to use the executor.
+It will act like so:
+
+1. Check if the runtime is running, if not spin up the runtime.
+2. Make a request to the runtime passing the arguments recieve through the API request.
+3. If the request is asyncronous then return the API request now with a execution ID like it does now.
+4. When we receive a response back from the runtime we will do 1 of two things
+    If the request is synchronous:
+        - Wait for the response from the runtime and return it as the response to the API request.
+    If the request is asynchronous:
+        - Store the result for later use through the Get Execution API endpoint. 
+
+`PATCH /v1/functions/{id}/tag` - Update function tag API endpoint
+The tag update endpoint will also recieve a small rewrite for the request moving it into the executor.
+
+It will act like so:
+1. Extract the code tarball into the enviroment.
+2. Check if the server is running, if it is then stop the server.
+3. also check if the runtime exists, if it doesn't then create the runtime.
+4. Map the new code into the runtime environment using a volume
+5. Finally bring the runtime back up and start the server again.
+
+#### Watchdog
+Watchdog will be apart of the executor running in the background dealing with all the running runtimes and monitoring their health.
```

Comment:
So how many processes will be running in the Executor container? Best Docker practice is to limit one process per container.

Reply:
Agree with @kodumbeats. This is a big overhead and probably not the best pattern for a Docker container.

Reply:
Watchdog could be split away into it's own process if that is the case :+1:

Reply:
I've removed watchdog from the RFC for now, so we can better focus on the core of the implementation